release/models/types/openconfig-yang-types.yang

```diff
@@ -124,10 +132,7 @@ module openconfig-yang-types {
       "A 32-bit counter. A counter value is a monotonically increasing
       value which is used to express a count of a number of
       occurrences of a particular event or entity. When the counter
-      reaches its maximum value, in this case 2^32-1, it wraps to 0.
-
-      Discontinuities in the counter are generally triggered only when
-      the counter is reset to zero.";
+      reaches its maximum value, in this case 2^32-1, it wraps to 0.";
```

Comment:
Given the context from @dplore, is it to say that if a command on the device such as "clear ip intf statistics counters" is issued it will clear only the counters used in native schema but not in OC?

Reply:
@hellt I'll give my viewpoint here since I know where you are going with this considering the wording `if OC counters are expected to be reset ...`

Maybe this should be conveyed somewhere more formally to set expectations correctly but I have yet to see this across any industry effort to date (I may just not have come across as well)

In _most_ cases we should clarify that there is a single state (I'll concentrate only on r/o for now) data-source for a given unique object in a given implementation and thus no per-schema or per-API datasets of sorts (I hope we all agree on this otherwise this warrants an entirely separate discussion).  This means if an implementation supports native, OpenConfig and say IETF modeling and there is a node that is meant to access the same underlying data, that this data is always represented consistently across all views (assuming type/representation matches and transformations aside)

Now, your example above is a bit unusual in intent so this opens up another case of concern in general.

First, there is no "OC" method (gNOI) to execute such an action - CLI is not a defined NBI for this world so this falls back to ymmv across implementation behaviors.

Second, the intent of this type of command across implementations is generally tied to a specific NBI which is the human interactive CLI (or even user session possibly which is what it really should be tied to at most imo) - a `clear` in this case does not mean to flush the counters on all NBI as you are not resetting IF-MIB OIDs, YANG state nodes, etc... but rather its use has been for real-time human debugging on the element directly that doesn't necessarily correspond to the paradigm of OC (at least not as it stands today) - a clear across all NBI as you can imagine could be more catastrophic affecting downstream systems consuming such data.

I feel like this topic probably does need to be recorded otherwise it will pop up from time to time on expected behaviors but as I see it any "clear counters" is local to the native implementation over CLI only

Reply:
@earies maybe an example would do good

if we have a leaf `ro some-counter` that in whatever schema contains a value `100`, then if we have a CLI show command `show some-counter` executed, it would return value 100.

Now if we execute `clear some-counter` in CLI, then the next `show some-counter` would return 0, but the underlying value for the leaf would still be `100`. Thus, `clear` in the CLI affects the representation of a value in the CLI, but does not nullify the underlying data.

Is that example matching what you described above, @earies?

Reply:
+1 to:  A "clear counters" CLI command should affect only the CLI view of the counters.     The operational use case for this CLI is debugging.  It's common to utilize the underlying hardware counters for systems (such as billing) to read the system counters and be negatively impacted by unexpected, repeated rollovers on counters.

Reply:
>A "clear counters" CLI command should affect only the CLI view of the counters. The operational use case for this CLI is debugging. It's common to utilize the underlying hardware counters for systems (such as billing) to read the system counters and be negatively impacted by unexpected, repeated rollovers on counters.

This is uncharted waters, and, practically speaking, will be unimplementable in many cases. The system will have to maintain N copies of every counter. But there are other situations where counter64 leafs are reset to zero. For example, a system reboot, usually, clears all counters to the default value. A component reboot (e.g. a linecard reboot, or a software component restart) may do the same, depending on the implementation. There could be other cases.

Also, w.r.t interface counters specifically, there's already some text about it in `openconfig-interfaces`, which allows clearing the counters.

Reply:
> This is uncharted waters

Agreed - we are taking an existing feature that contradicts the management and debuggability goals under this umbrella imo

> practically speaking, will be unimplementable in many cases

Well, various implementations have supported what I describe for years already.  A `clear counters` is _only_ segmented to the CLI NBI - Arista EOS and JUNOS/EVO work this way today (I cannot recall other implementations off the top of my head)

```
eos1#show interfaces Ethernet 1 | include clearing|packets.*put
  Last clearing of "show interface" counters never
     544 packets input, 58805 bytes
     465 packets output, 59009 bytes

interfaces/interface[name=Ethernet1]/state/counters/in-octets: 67993

IF-MIB::ifInOctets.1 = Counter32: 67659

eos1#clear counters ethernet 1
eos1#show interfaces Ethernet 1 | include clearing|packets.*put
  Last clearing of "show interface" counters 0:00:23 ago
     10 packets input, 814 bytes
     13 packets output, 1641 bytes

interfaces/interface[name=Ethernet1]/state/counters/in-octets: 68400

IF-MIB::ifInOctets.1 = Counter32: 69068
```

> there's already some text about it in openconfig-interfaces, which allows clearing the counters.

For context, here is the text for `in-octets` today

```
description
  "The total number of octets received on the interface,
  including framing characters.

  Discontinuities in the value of this counter can occur
  at re-initialization of the management system, and at
  other times as indicated by the value of
  'last-clear'.";
reference
```

I think this text is incorrect and implies that the leafs values are cleared in correspondence to the `last-clear` leaf which is not true in various implementations.  This leaf in those implementations is not cleared by any forceful "clear" command but would be if there is a process/hw reinit

@LimeHat @hellt - in SRLinux/SROS does this correspond to the behavior of your native "clear counters"?

> if we have a leaf ro some-counter that in whatever schema contains a value 100, then if we have a CLI show command show some-counter executed, it would return value 100.
>
> Now if we execute clear some-counter in CLI, then the next show some-counter would return 0, but the underlying value for the leaf would still be 100. Thus, clear in the CLI affects the representation of a value in the CLI, but does not nullify the underlying data.
>
> Is that example matching what you described above, @earies?

Yep you got it...  The leaf `last-clear` imo is only of value to the CLI user to understand why show commands are displaying different than the actual underlying data-source (my observation across a few implementations is this is across the entirety of the CLI NBI and not per session)

Reply:
>in SRLinux/SROS does this correspond to the behavior of your native "clear counters"?

We follow a single source of truth approach and the cli view will match gnmi/netconf/any other interface's view.

Reply:
> > in SRLinux/SROS does this correspond to the behavior of your native "clear counters"?
> 
> We follow a single source of truth approach and the cli view will match gnmi/netconf/any other interface's view.

With this logic, this could really be applied generically to "reset" any state counter in the system at the source-of-truth.... nothing specific to interface counters then for what was a historical human-debugging use-case.

In either case a consumer needs to handle rollover and this approach just means a client triggered action can force.

Another implementation observation (IOS-XR): A clear will reset the state leaf for YANG based access (gNMI/NETCONF) + CLI but not for SNMP IF-MIB.

I know this conversation is going down a bit of a divergent path of pointing out some behavioral differences but it seems that the usefulness of a `last-clear` and user-defined triggers may very well warrant common behaviors which I tend to agree about maintaining and operating on a single copy/source-of-truth.

This could warrant a `last-clear` node at _every_ `./counters` container + generic methods to clear any counter within a system.  A generic gNOI RPC could also cover this by way of targeting any suffixed path of `../../counters`.

This would introduce backwards incompatible yet consistent behaviors across various implementations

Reply:
I have a couple of thoughts on this thread:

Firstly, I would argue that specifying the behaviour of native CLI commands is somewhat out of scope for OpenConfig specifications, even though I see benefits for operators if systems behave similarly.

But generally, we think that it makes sense for the CLI to clear the counters everywhere, otherwise, this means that the device needs to burn extra memory keeping a separate copy of the counters for CLI vs OC vs IETF vs MIBs, etc.

I can see the argument about wanting to debug network issues via the CLI, but I think that could be better achieved by a separate transient command that takes a snapshot of the counters that could then be compared with the values at a subsequent point entirely independently from the main stats collection.  Then, when the snapshot is no longer required, the associated resources can be freed.